### PR TITLE
Fallback to UTF8 converting command line arguments

### PR DESCRIPTION
**Pull request by pshipton**
On Linux the encoding of the terminal seems to be unknown and isn't necessarily `locale charmap`. When converting command line arguments and the conversion fails using the default encoding, try UTF8 as well.

OpenJDK accepts UTF8 input.

A way to test this is to have a locale such as `zh_CN.gb18030` installed, and then run `LANG=zh_CN.gb18030 java -Dzz=〢€郎嗀秊㐀龦 -XshowSettings:properties -version`. The zz= characters should be UTF8 encoded. In a terminal using UTF8, and without this change the JVM fails to start with `Error: Could not create the Java Virtual Machine.`.

### CODE REVIEW

File: runtime/vm/vmprops.c [hunk 1]
```diff
@@ -1316,6 +1316,28 @@ containsBackslashU(const char *userString, UDATA stringLength) {
 	return FALSE; /* hit the end of the string without encountering \u sequence */
 }
 
+static U_8 *
+convertString(J9JavaVM *vm, I_32 fromCode, const char *userString, UDATA stringLength) {
```
Comment (keithc-ca) [inline runtime/vm/vmprops.c:1320]: Brace (`{`) on next line, please.

File: runtime/vm/vmprops.c [hunk 2]
```diff
@@ -1316,6 +1316,28 @@ containsBackslashU(const char *userString, UDATA stringLength) {
...
+	PORT_ACCESS_FROM_JAVAVM(vm);
+	U_8 *result = NULL;
+	I_32 bufferLength = j9str_convert(fromCode, J9STR_CODE_MUTF8, userString, stringLength, NULL, 0);
+	if (bufferLength >= 0) {
+		U_8 *mutf8Buffer = NULL;
+		++bufferLength; /* account for terminating null */
+		mutf8Buffer = j9mem_allocate_memory(bufferLength, OMRMEM_CATEGORY_VM);
```
Comment (keithc-ca) [inline runtime/vm/vmprops.c:1327]: Suggest accounting for the NUL earlier and adjusting the condition:
```c
	I_32 bufferLength = j9str_convert(fromCode, J9STR_CODE_MUTF8, userString, stringLength, NULL, 0) + 1; /* +1 for terminating null */
	if (bufferLength > 0) {
		U_8 *mutf8Buffer = j9mem_allocate_memory(bufferLength, OMRMEM_CATEGORY_VM);
```

File: runtime/vm/vmprops.c [hunk 3]
```diff
@@ -1316,6 +1316,28 @@ containsBackslashU(const char *userString, UDATA stringLength) {
...
+		if (NULL != mutf8Buffer) {
+			I_32 resultLength = j9str_convert(fromCode, J9STR_CODE_MUTF8, userString, stringLength,
+					(char*)mutf8Buffer, bufferLength);
```
Comment (keithc-ca) [inline runtime/vm/vmprops.c:1330]: Space before `*`, please.

File: runtime/vm/vmprops.c [hunk 4]
```diff
@@ -1316,6 +1316,28 @@ containsBackslashU(const char *userString, UDATA stringLength) {
...
+			} else {
+				j9mem_free_memory(mutf8Buffer);
+			}
+		}
+	}
+	return result;
+}
 
 U_8*
 getMUtf8String(J9JavaVM * vm, const char *userString, UDATA stringLength) {
```
Comment (keithc-ca) [inline runtime/vm/vmprops.c:1343]: Space before `*`, `{` on a separate line:
```c
U_8 *
getMUtf8String(J9JavaVM *vm, const char *userString, UDATA stringLength)
{
```

**Review by keithc-ca — CHANGES_REQUESTED** [review]

File: runtime/vm/vmprops.c (hunk 4 above)
Comment (pshipton) [reply runtime/vm/vmprops.c]: Pushed the changes to address these review comments.

File: runtime/vm/vmprops.c [hunk 5]
```diff
@@ -1316,10 +1316,31 @@ containsBackslashU(const char *userString, UDATA stringLength) {
...
 }
 
+static U_8 *
+convertString(J9JavaVM *vm, I_32 fromCode, const char *userString, UDATA stringLength)
+{
+	PORT_ACCESS_FROM_JAVAVM(vm);
+	U_8 *result = NULL;
+	I_32 bufferLength = j9str_convert(fromCode, J9STR_CODE_MUTF8, userString, stringLength, NULL, 0) + 1; /* +1 for terminating null */
+	if (bufferLength > 0) {
+		U_8 *mutf8Buffer = mutf8Buffer = j9mem_allocate_memory(bufferLength, OMRMEM_CATEGORY_VM);
```
Comment (keithc-ca) [inline runtime/vm/vmprops.c:1326]: `mutf8Buffer = mutf8Buffer` -> `mutf8Buffer`

**Review by keithc-ca — CHANGES_REQUESTED** [review]

File: runtime/vm/vmprops.c (hunk 5 above)
Comment (pshipton) [reply runtime/vm/vmprops.c]: Updated

**Comment by keithc-ca** [comment]
Jenkins compile amac jdk21

**Review by keithc-ca — APPROVED** [review]

**Comment by keithc-ca** [comment]
Verified locally that the change has the described effect.

**Pull request merged**